### PR TITLE
Fix lgw flows for ingress-svc traffic

### PR DESCRIPTION
**Pull request by tssurya**
This commit does two things:

1) No need to steer service traffic for LGW via breth0 into OVN.
Let's pass it into the host where user can apply custom routes
and modulate the traffic as needed.
Flow: external->br-ex->host->iptables->br-ex->GR->pod
reply will follow the same path (check docs for details)
2) Cleanup some ETP=local code since this simplifies things on
br-ex for LGW a lot!

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by tssurya** [comment]
/retest
failure due to https://github.com/ovn-org/ovn-kubernetes/issues/2846#issuecomment-1060394042

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 1]
```diff
@@ -75,30 +75,48 @@ type serviceConfig struct {
...
+		} else {
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=%s:%s),table=6)",
+					cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv4, svcPort.TargetPort.String()))
+		}
+		nodeportFlows = append(nodeportFlows,
+			// table 0, Matches on return traffic, i.e traffic coming from the host networked pod's port, and unDNATs
+			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, tp_src=%s, actions=ct(commit,zone=%d nat,table=7)",
+				cookie, flowProtocol, svcPort.TargetPort.String(), HostNodePortCTZone))
 	} else {
```
Comment (trozet) [inline go-controller/pkg/node/gateway_shared_intf.go:101]: wait, cant a service have external IP, LB ingress IP, and also be nodeport? why is this else here? @astoycos ?

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 2]
```diff
@@ -292,74 +279,44 @@ func (npw *nodePortWatcher) createLbAndExternalSvcFlows(service *kapi.Service, s
...
 	}
+	npw.ofm.updateFlowCacheEntry(key, externalIPFlows)
 
 	return nil
 }
 
 // generate ARP/NS bypass flow which will send the ARP/NS request everywhere *but* to OVN
 // OpenFlow will not do hairpin switching, so we can safely add the origin port to the list of ports, too
-func (npw *nodePortWatcher) generateArpBypassFlow(protocol string, ipAddr string, cookie string) string {
+func (npw *nodePortWatcher) generateArpBypassFlow(protocol string, ipAddr string, cookie string) []string {
```
Comment (trozet) [inline go-controller/pkg/node/gateway_shared_intf.go:301]: why is this changed to return a slice when the logic of the function hasn't changed to return more than 1 value?

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 3]
```diff
@@ -75,30 +75,48 @@ type serviceConfig struct {
...
-	nodeportFlowsResult = append(nodeportFlowsResult,
+		// table 0, This rule matches on all traffic with dst port == NodePort, DNAT's it to the correct NodeIP
+		if strings.Contains(flowProtocol, "6") {
+			// If ipv6 make sure to choose the ipv6 node address for rule
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=[%s]:%s),table=6)",
+					cookie, npw.ofportPhys, flowProtocol, svcPort.NodePort, HostNodePortCTZone, npw.gatewayIPv6, svcPort.TargetPort.String()))
+		} else {
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=%s:%s),table=6)",
```
Comment (trozet) [inline go-controller/pkg/node/gateway_shared_intf.go:111]: we are missing comments on what these flows are doing

**Review by trozet — CHANGES_REQUESTED** [review]

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_shared_intf.go:101]: oh I think this is the semantic of the function, if externalIPOrLBIngressIP is empty, then assume its nodeport. I think we should change this function and re-examine this code. This now seems overly complicated and not very clear.

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 4]
```diff
@@ -75,30 +75,48 @@ type serviceConfig struct {
...
-		nodeportFlowsResult = append(nodeportFlowsResult,
-			// Traffic destined for ETP=local backed by OVN-K pod in LGW mode is matched on the svcNP and sent to table6.
-			fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,table=6)",
-				cookie, npw.ofportPhys, flowProtocol, svcPort, HostNodePortCTZone),
-			// Return traffic is matched on the svcNP as the sourcePort and gets un-DNATed back to nodeIP.
-			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, tp_src=%d, actions=ct(zone=%d,table=7)",
-				cookie, flowProtocol, svcPort, HostNodePortCTZone))
-	}
-	nodeportFlowsResult = append(nodeportFlowsResult,
+		// table 0, This rule matches on all traffic with dst port == NodePort, DNAT's it to the correct NodeIP
```
Comment (trozet) [inline go-controller/pkg/node/gateway_shared_intf.go:103]: if its nodeport isnt the destination IP already the node ip? Looks like thsis is DNAT'ing the node port -> cluster target port?

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 5]
```diff
@@ -246,13 +232,12 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add, h
 // case1: If a service has externalTrafficPolicy=local, and has host-networked endpoints local to the node (hasLocalHostNetworkEp),
 // traffic instead will be steered directly into the host and DNAT-ed to the targetPort on the host.
 //
-// case2: Only applicable for LGW mode: If a service has externalTrafficPolicy=local, and it doesn't have host-networked
-// endpoints local to the node (!hasLocalHostNetworkEp - including empty endpoint.Subset), traffic will be steered into LOCAL
-// preserving sourceIP and IPTables will steer this traffic into OVN via ovn-k8s-mp0.
-//
-// case3: All other types of services i.e:
+// case2: All other types of services i.e:
 //        case3a: if externalTrafficPolicy=cluster, irrespective of gateway modes, traffic will be steered into OVN via GR.
```
Comment (trozet) [inline go-controller/pkg/node/gateway_shared_intf.go:236]: this isn't true anymore right?

File: go-controller/pkg/node/gateway_shared_intf.go [hunk 6]
```diff
@@ -292,74 +279,44 @@ func (npw *nodePortWatcher) createLbAndExternalSvcFlows(service *kapi.Service, s
...
 		// case2 (see function description for details)
-		var nodeportFlows []string
-		klog.V(5).Infof("Adding flows on breth0 for %s Service %s in Namespace: %s since ExternalTrafficPolicy=local", ipType, service.Name, service.Namespace)
-		nodeportFlows = npw.createFlowsForNonLocalHostNetEp(cookie, svcPort.Port, flowProtocol, externalIPOrLBIngressIP, nwDst, nwSrc)
-		npw.ofm.updateFlowCacheEntry(key, nodeportFlows)
-	} else {
-		// case3 (see function description for details)
-		npw.ofm.updateFlowCacheEntry(key, []string{
+		externalIPFlows = append(externalIPFlows,
 			fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, "+
```
Comment (trozet) [inline go-controller/pkg/node/gateway_shared_intf.go:287]: lets put comments on these flows saying what they are doing.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 2 above)
Comment (tssurya) [reply go-controller/pkg/node/gateway_shared_intf.go:301]: haha ok my bad, I was lazy so changed return value to hack my need to append the results. let me change this back and do the needful at the right place in code.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 4 above)
Comment (tssurya) [reply go-controller/pkg/node/gateway_shared_intf.go:103]: yes precisely, there is no change to IP; only the target port changes.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 3 above)
Comment (tssurya) [reply go-controller/pkg/node/gateway_shared_intf.go:111]: its on L103, let me move it down.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 5 above)
Comment (tssurya) [reply go-controller/pkg/node/gateway_shared_intf.go:236]: yeah good catch again, let me update the comment.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 6 above)
Comment (tssurya) [reply go-controller/pkg/node/gateway_shared_intf.go:287]: done.

**Comment by tssurya** [comment]
opened https://github.com/ovn-org/ovn-kubernetes/issues/2868 for the e2e dual stack conversion failure.

File: go-controller/pkg/node/gateway_shared_intf.go (hunk 1 above)
Comment (tssurya) [reply go-controller/pkg/node/gateway_shared_intf.go:101]: @trozet maybe I should not modularize this :D I think that will be more clear, wdyt?

Comment (tssurya) [reply go-controller/pkg/node/gateway_shared_intf.go]: done.

**Comment by tssurya** [comment]
hmm https://github.com/ovn-org/ovn-kubernetes/runs/5738406898?check_suite_focus=true#step:14:14427 the dual-conversion job is failing on a couple of PRs, no idea why let me run them locally.

**Comment by tssurya** [comment]
ok its not just this PR, seems like its flaking on another PR as well: https://github.com/ovn-org/ovn-kubernetes/runs/5682105064?check_suite_focus=true#step:14:11636

File: docs/design/service_traffic_policy.md [hunk 7]
```diff
@@ -219,9 +218,100 @@ to DNAT directly to the correct host endpoint.
...
+name                : "Service_default/example-service-1_TCP_cluster"
+options             : {event="false", reject="true", skip_snat="false"}
+protocol            : tcp
+selection_fields    : []
+vips                : {"10.96.67.170:80"="10.244.1.3:8080,10.244.2.3:8080"}
+```
+
+Response traffic will follow the same path (backend->GR->breth0->host->breth0->eth0).
+
+7. Return traffic gets matched on the priprity 500 flow in `table0` which sends it to `table3`.
```
Comment (squeed) [inline docs/design/service_traffic_policy.md:296]: priority :)

**Comment by squeed** [comment]
This is a bit tricky to grok, but seems good to me.

File: docs/design/service_traffic_policy.md (hunk 7 above)
Comment (tssurya) [reply docs/design/service_traffic_policy.md:296]: oh oops, let me fix this.

**Comment by tssurya** [comment]
lint failure should be fixed by https://github.com/ovn-org/ovn-kubernetes/pull/2889
and looking into e2e dul stack failure on LGW, not sure if its my PR or the test (https://github.com/ovn-org/ovn-kubernetes/issues/2868) at this point

**Comment by tssurya** [comment]
spoke to @andreaskaris : doesn't seem like my PR, he is investigating that failure via https://github.com/ovn-org/ovn-kubernetes/pull/2869

**Review by trozet — APPROVED** [review]
